api/token.go

```diff
@@ -63,6 +65,58 @@ func (a *API) ResourceOwnerPasswordGrant(ctx context.Context, w http.ResponseWri
 	a.issueRefreshToken(user, w)
 }
 
+// AuthorizationCodeGrant implements the authorization_code grant for use with external OAuth providers
+func (a *API) AuthorizationCodeGrant(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	code := r.FormValue("code")
+	providerName := r.FormValue("provider")
+
+	if code == "" {
+		log.Printf("No authorization code found: %v", r)
+		sendJSON(w, 400, &OAuthError{Error: "invalid_request", Description: "Authorization code required"})
+		return
+	} else if providerName == "" {
+		log.Printf("No provider name found: %v", r)
+		sendJSON(w, 400, &OAuthError{Error: "invalid_request", Description: "External provider name required"})
+		return
+	}
+
+	provider, err := a.Provider(providerName)
+	if err != nil {
+		log.Printf("Unsupported provider: %+v", err)
```

Comment:
instead of this it might make more sense to use a log that is a `logrus.Entry` on the API itself. That way that logging is handled in the rest of the project. 

And if you're just repeating the error it doesn't seem like you need that.